Data models

kappq opened the pull request:
Add the models for the event types.

It contains:
- An `EventType` enum, represents all the event types.
- An `EventData` class, inherits from `pydantic.BaseModel`.
- A class for every type of event, inherits from `EventData` and specifies the required data for the type of event.
- An `Event` class, puts together the event type and data.

kappq committed:
Add models for event types

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
```
Isn't it `from __future__ import __annotations__`?

LukeHankey replied:
Nope, nvm. I thought I did that in my work repo so had to double check :p

kappq replied:
This gives a `SyntaxError`:
```
SyntaxError: future feature __annotations__ is not defined
```

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,98 @@
+        collaborators: The list of users that already collaborate in the room.
+    """
+
+    code: str
+    collaborators: list[str]
```
Does this not need the `room_code` as well?

kappq replied:
It shouldn't be needed. The `room_code` is obtained with the initial `connection` event, otherwise, every event would need the room code.

But now that you make me think about that, it probably isn't even needed in the `DisconnectData`.

LukeHankey replied:
Yeah, I agree. Don't think it needs to be in `DisconnectData`.

kappq replied:
But it probably needs `username`, so that the client knows which collaborator to remove from the list.

LukeHankey replied:
Agreed.

LukeHankey commented:
What about the status_code? Websockets can have custom codes in the 4000 range. We currently send 1 error back to the client  when they join a room that doesn't exist

kappq committed:
Remove room_code from DisconnectData

kappq committed:
Add username to DisconnectData

LukeHankey commented:
Do you want to remove:
```py
class ConnectionEventData(BaseModel):
    """A model representing the connection event data."""

    message: str
    difficulty: int
    room_code: str
    connection_type: Literal["create", "join"]


class ConnectionData(BaseModel):
    """A model representing the initial connection."""

    type: str
    data: ConnectionEventData
```
from main.py in this PR or create another for that?

kappq commented:
> Do you want to remove:
> 
> ```python
> class ConnectionEventData(BaseModel):
>     """A model representing the connection event data."""
> 
>     message: str
>     difficulty: int
>     room_code: str
>     connection_type: Literal["create", "join"]
> 
> 
> class ConnectionData(BaseModel):
>     """A model representing the initial connection."""
> 
>     type: str
>     data: ConnectionEventData
> ```
> 
> from main.py in this PR or create another for that?

We might as well remove it here.

kappq commented:
> What about the status_code? Websockets can have custom codes in the 4000 range. We currently send 1 error back to the client when they join a room that doesn't exist

I made another PR for this, checkout #42 .

kappq committed:
Add error event type

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,109 @@
+    """
+
+    type: EventType
+    data: EventData
```
Add a `status_code: StatusCode` when #42 is merged

kappq replied:
Fixed in 2b7443fc0ba2fb828df397ae96eee22a1143283c.

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,109 @@
+    """
+
+    type: EventType
+    data: EventData
```
Add this to the Event class and fill in the rest of the cases

```py
    @validator("data", pre=True)
    def valid_data(cls, value, values):
        match values["type"]:
            case EventType.CONNECT:
                value = ConnectionData(**value)
        return value
```
Explanation:
- `value`: The value from the validator (data)
- `values`: The previous values as a dictionary
- `pre` in the validator decorator means use this validation first before pydantics built-in validation which allows us to get the actual values from the **await client.receive() as `value`

kappq replied:
Fixed in 81a8a8f970dceaf3061b5522d1428a2284f2c593.

kappq committed:
Rename ConnectionData to ConnectData

kappq committed:
Add validator to Event

kappq committed:
Merge branch 'main' into data-models

kappq committed:
Add status_code to Event

kappq committed:
Remove old event data classes

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +202,44 @@ async def room(websocket: WebSocket) -> None:
+        return
+
+    initial_data: ConnectData = initial_event.data
+    print(initial_event, initial_data)
```
Left in prints :)

kappq replied:
Fixed in 887a4b63a66dfa1e0df0dcf0ee93ee715da73394.

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +202,44 @@ async def room(websocket: WebSocket) -> None:
+    print(initial_event, initial_data)
+    room_code = initial_data.room_code
+
+    match initial_data.connection_type:
+        case "create":
+            manager.create_room(client, room_code)
+        case "join":
+            try:
+                manager.join_room(client, room_code)
+            except RoomNotFoundError as e:
+                await client.send(
+                    Event(type=EventType.ERROR, data=ErrorData(message=e.message), status_code=StatusCode.SUCCESS)
+                )
+                await client.close()
+                return
+
+    manager.broadcast(initial_event, room_code)
```
Can probably recreate the ConnectionManager.connect method to abstract this away

kappq replied:
I probably wouldn't leave the `ConnectionManager` the job of handling events.

LukeHankey replied:
This would only be for connecting the client to the room though, seems fitting it would go in the `ConnectionManager`, no? Unless you had another idea.

Event handling will mostly be between receiving the data and broadcasting in the `while True:` loop. So it would be either in the client, manager, or another class.

kappq replied:
Fixed in aff01c8f0c7d4c14ef2df3a682e569d16ea1e8ca. But I left the error handling for `RoomNotFoundError` in the `room` endpoint.

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,128 @@
+    status_code: StatusCode
+
+    @validator("data", pre=True)
+    def valid_data(self, value, values):
```
Should be `cls`, not `self`.

kappq replied:
Fixed in 2a37e8232350db2196d11c144f6370ad3dbabfe0. `flake8` was complaining about that so I also ignored the warning.

kappq committed:
Rename cls to self

Configures flake8 to ignore N805.

kappq committed:
Remove useless print statement

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +202,44 @@ async def room(websocket: WebSocket) -> None:
 
     try:
         while True:
             data = await client.receive()
```
`data = Event(**await client.receive())`

kappq replied:
Fixed in bdd6848c260861f606d0fe7f3d885646d55fa8e8.

kappq committed:
Convert received data to Event

kappq committed:
Move ConnectionManager methods

kappq committed:
Add ConnectionManager.connect method

This method handles the connection type directly in the ConnectionManager.

kappq committed:
Convert ConnectionManager.connect to a static method

kappq committed:
Make _update_code_cache public

This also sets the line length for isort to 119.

kappq committed:
Make EventType a subclass of str

This helps with JSON serialization.

kappq committed:
Convert data type to Event

This also changes the docstrings of the modified functions.

kappq committed:
Update code cache tests

Updates the tests because _update_code_cache was made public.

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,131 @@
+        username: The username of the user creating or joining the room.
+    """
+
+    connection_type: Literal["create", "join"]
+    room_code: str
+    difficulty: int | None
+    username: str
```
Match the order with the Fields in docstring

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +214,44 @@ async def room(websocket: WebSocket) -> None:
+    except RoomNotFoundError as e:
+        await client.send(
+            Event(type=EventType.ERROR, data=ErrorData(message=e.message), status_code=StatusCode.ROOM_NOT_FOUND)
+        )
```
Could we add this to the error itself to format in a data attribute and then we can just do `await client.send(e.data)`

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,131 @@
+
+    type: EventType
+    data: EventData
+    status_code: StatusCode
```
We shouldn't be receiving status_codes from the client, only sending them

kappq replied:
> We shouldn't be receiving status_codes from the client, only sending them

Do I just make it optional?

LukeHankey replied:
Rename Event to `EventRequest` and add another event for Response.
```py
class EventResponse(EventRequest):
    status_code: StatusCode
```

kappq replied:
Fixed in d5d33ecc36fbfeb4490f1acee3e48b21e136a19c.

kappq committed:
Match field order with docstring

kappq committed:
Add data attribute to RoomNotFoundError

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +202,44 @@ async def room(websocket: WebSocket) -> None:
+                await client.close()
+                return
+
+    manager.broadcast(initial_event, room_code)
```
Why do we need the broadcast here? It's currently sending the same request back upon creating. When using `join`, it breaks

kappq replied:
To let other clients know when a new user connects to the room, this won't be needed when `sync` is implemented.

LukeHankey commented on server/errors.py:
```diff
@@ -7,5 +8,4 @@ class RoomNotFoundError(Exception):
 
-        self.code = StatusCode.ROOM_NOT_FOUND
-        self.message = message
+        self.data = Event(type=EventType.ERROR, data=ErrorData(message=message), status_code=StatusCode.ROOM_NOT_FOUND)
```
`data=ErrorData(message=message).dict()` or pass in the data as a dictionary with the message key. Perhaps it would be easier to do:
```py
self.message = {"message": message}
self.data = Event(type=EventType.ERROR, data=message, status_code=StatusCode.ROOM_NOT_FOUND)
```

The data in Event must be a mapping to use the `EventData(**value)` in events.py

kappq replied:
Fixed in 83f16e4997d25b12d4f4c162a66d293ae8f6ed6a.

bensgilbert commented on server/events.py:
```diff
@@ -0,0 +1,131 @@
+    """
+
+    connection_type: Literal["create", "join"]
+    difficulty: int | None
```
shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

LukeHankey replied:
> shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

`Optional` is essentially an alias for `Union[thing, None]`. In this case it would be `Union[int, None]` which in Python 3.10 syntax is `int | None`

kappq replied:
> shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

`Optional` is equal to `Type | None`. `pydantic` is smart enough to know that optional fields are not required.

If you try to send a connection event without `difficulty` it works fine.

That said, we might need some more validators to check if we have the `difficulty` when the `connection_type` is `create`.

kappq committed:
Edit Event validator to accept EventData

LukeHankey commented on server/events.py:
```diff
@@ -115,8 +115,11 @@ class Event(BaseModel):
 
     @validator("data", pre=True)
-    def valid_data(cls, value, values):
+    def valid_data(cls, value: EventData | Mapping, values):
```
`Type[EventData] | Mapping` The left hand side shows that it should be a subclass of EventData

kappq replied:
We won't receive the object's type, but the object itself though.

LukeHankey replied:
Nvm, don't do this :p

kappq committed:
Add difficulty validator

kappq committed:
Separate EventRequest and EventResponse

EventRequest is the data sent from the client to the server, EventResponse is the data sent from the server to the client. It allows to easily handle the status codes.

LukeHankey commented on server/events.py:
```diff
@@ -102,17 +109,20 @@ class ErrorData(EventData):
     Fields:
         type: The type of the event.
         data: The data of the event.
+        status_code (optional): The status code of the event. Only required when
+            sending an event from the server to the client.
```
Don't need in EventRequest

kappq replied:
Fixed in 87edde6328f2a6ec25169ab20d22a8747c4eccc1.

kappq committed:
Fix difficulty type annotation

LukeHankey commented on server/events.py:
```diff
@@ -0,0 +1,159 @@
+    This represent a response sent from the server to the client.
+
+    Fields:
+        type: The type of the event.
+        data: The data of the event.
```
Does this need to be included?

kappq replied:
Yeah, not really needed, I included it just for consistency.

kappq replied:
Fixed in 923ec688561195ee0fbb055adec4ea28d2e67473.

kappq committed:
Fix EventRequest and EventResponse docstrings

kappq committed:
Remove useless docstrings

LukeHankey commented on server/main.py:
```diff
@@ -218,38 +214,42 @@ async def room(websocket: WebSocket) -> None:
+                data=DisconnectData(username=initial_data.username),
+                status_code=StatusCode.SUCCESS,
+            ),
+            room_code,
+        )
```
```py
    room_code,
    sender=cleint
)
```
This fixes the RuntimeError so that it doesn't broadcast a message to itself if they have already disconnected

kappq replied:
Fixed in 3b9af74630826320c39bdfe2a6b73d31a2f7b25d.

kappq committed:
Fix RuntimeError

This avoids that the client sends a message to itself after he disconnects.
